iOS 并发与多线程完整指南 (#9354)
xitu/gold-miner: article/2022/the-complete-guide-to-concurrency-and-multithreading-in-ios.md

Async / Await / Actor Is iOS13+
```

本文设计较多主题。如果有些内容你已经很熟悉了，可以尝试跳过它，阅读你还未熟悉的部分。会有一些技巧和提示。

## 现实世界中的多线程实例

假设你有一个餐厅。服务员正在收集订单。厨房在准备食物，酒保在制作咖啡和鸡尾酒。

> 在某些时候，很多人都会点咖啡和食物。这需要更多的时间来准备。此时钟声突然响起，5份食物和4份咖啡已经准备完成。但是，即使所有的产品都已备齐，服务员也需要逐一送到不同的餐桌上。这就是**串行队列**。使用串行队列，你只能有一个服务员。

Reviewer: ```suggestion
有时，会有很多人点咖啡和食物这些需要更多时间来准备的东西。此时铃声突然响起，5 份食物和 4 份咖啡已经准备完成。但是，即使所有的产品都已备齐，服务员也要逐一送到不同的餐桌上。这就是**串行队列**。使用串行队列，你只能有一个服务员。
```

现在假设有两个或三个服务员。他们可以在同一时间以更快的速度为各桌服务。这就是**并行**。即使用多个CPU来操作多个进程。

> 接下来继续假设一下，一个服务员不会同时为一个桌子服务，而是先为所有桌子的人送上所有的咖啡。然后，他们会询问一些新桌子的订单，然后为所有的食物服务。这个概念被称为**并发**。即在同一时间段内进行上下文切换，管理以及多计算任务。这并不一定意味着它们会在同一时刻都在运行。例如，在一个单核机器上进行多任务。

Reviewer: 「它是上下文切换、管理以及同时进行的多计算任务」这句感觉不太通顺
```suggestion
继续假设，一个服务员不只服务一张桌子，而是先为所有桌子的顾客送上所有的咖啡，然后，他们会为一些新来的顾客进行点单，最后才供应所有的食物。这个概念被称为**并发**。即在同一时间段内进行上下文切换，管理以及多计算任务。这并不一定意味着它们会在同一时刻都在运行。例如，在单核机器上进行的多任务处理。
```

如今的设备都有多个CPU（[中央处理器](https://en.wikipedia.org/wiki/Central_processing_unit)）。为了能够创建具有无缝流转的应用程序，我们需要了解**多线程**的概念。这是应用程序中处理多任务的一种方式。很重要的一点是，我们需要明白，如果某个东西 "有效"，它也许不是最好的、最理想的方式。很多时候，我看到一个长期运行的任务发生在UI线程上，并且阻塞了应用程序几秒钟。现在，对于用户来说，这可能是一个抉择的时刻。用户可能会删除你的应用程序，因为他们觉得其他的应用程序启动得更快，获取书籍、下载音乐的速度更快。竞争是巨大的，人们都追求着更优秀的体验。

![Thread execution types](https://cdn-images-1.medium.com/max/2000/1*VAGYSwgHGTHvAPhP1m204Q.png)

> 你可以看到 "第六节 "任务，如果它是在当前时间安排的。它将以 **FIFO** 的方式添加到列表中，并等待在未来执行。

Reviewer: Serial 6 不需要翻译成第六节，和图片对应就好
```suggestion
如果 "Serial 6" 的任务是在当前时间安排的，那么你就可以看到它了。它将以 **FIFO** 的方式添加到列表中，并等待将来的执行。
```

## 多线程的基础

> 我一直纠结的一件事是，没有标准的术语。为了帮助解决这个问题，对于这些主题，我将首先写下同义词、例子。如果你使用iOS以外的其他技术，你仍然可以理解这个概念并将其映射，因为基本原理是相同的。幸运的是，在我职业生涯的早期，我一直在使用C、C++、C#、node.js、Java（Android）等技术栈，所以我已经习惯了这种上下文切换。

Reviewer: 转移 => 运用到其他技术上
```suggestion
我一直纠结的一件事是，没有标准的术语。为了帮助解决这个问题，对于这些主题，我将首先写下同义词、例子。如果你使用 iOS 以外的其他技术，你仍然可以理解这个概念并将其运用到其他技术上，因为基本原理是相同的。幸运的是，在我职业生涯的早期，我一直在使用 C、C++、C#、node.js、Java（Android）等技术栈，所以我已经习惯了这种上下文的切换。
```

}
```

> * **后台线程（全局）:** 预定义的。大多数情况下，我们根据自己的需要在新线程上创建任务。例如，如果我们需要下载体积较大的图片。这是在后台线程上完成，或者调用一些API。我们不想因为等待这个任务的完成而阻止用户的操作。我们将在后台线程上调用一个API来获取一个电影数据的列表。当收到响应并完成数据解析后，我们便切换至主线程上并更新用户界面。

Reviewer: ```suggestion
* **后台线程（全局）:** 预定义的。大多数情况下，我们根据自己的需要在新线程上创建任务。例如，如果我们需要下载体积较大的图片，这是在后台线程上完成的。又例如当调用一些 API时，我们不想因为等待这个任务的完成而阻止用户的操作。我们将在后台线程上调用一个 API 来获取一个电影数据的列表。当收到响应并完成数据解析后，我们便切换至主线程上并更新用户界面。
```

let concurQueue = DispatchQueue(label: “com.kraken.concurrent”, attributes: .concurrent)
```

我们还可以创建一个私有队列，它也可以是串行和并发的。

## GCD (Grand Central Dispatch)

GCD是苹果的低级线程接口，用于支持多核硬件上的并发代码执行。GCD以一种简单的方式，让你的手机可以在后台下载视频的同时保持用户界面的响应。

> "DispatchQueue是一个管理应用程序的主线程或后台线程任务的串行或并发执行的对象。" —— Apple Developer

如果你注意到上面的代码示例，你可以看到 "qos"。这指的是服务质量。通过这个参数，我们可以定义如下的优先级。

* **background** — 当一个任务对时间不敏感，或者当用户可以在这个过程中做一些其他的互动时，我们可以使用这个方法。比如预先获取一些图片，加载，或者在这个后台处理一些数据。这项工作需要大量的时间，秒、分钟、小时。
* **utility** — 长期运行的任务。一些处理用户可以看到的东西。例如，下载一些带有指标的地图。当一个任务需要几秒钟，最终需要几十分钟。

> * **userInitiated** — 当用户从用户界面启动一些任务并等待结果以继续与应用程序互动时。这个任务需要几秒钟或一瞬间。

Reviewer: ```suggestion
* **userInitiated** — 用户从用户界面启动一些任务并等待结果以继续与应用程序交互。这个任务需要几秒钟或一瞬间。
```

> * **userInteractive** — 当用户需要立即完成某些任务，以便能够继续与应用程序进行下一次互动。即时任务。

Reviewer: ```suggestion
* **userInteractive** — 用户需要立即完成某些任务，以便能够继续与应用程序进行下一次交互。是一个即时任务。
```

}
```

> 我的观点是，就异常而言，这是万恶之源。对于每一个需要延迟的异步任务，我建议要考虑清楚，如果可能的话，可以使用一些状态管理。不要把这个选项作为第一选择。通常情况下，会有其他解法。

Reviewer 2: with respect for exceptions
这个的翻译需要斟酌一下，这里建议直接省略掉，也不影响所表达的意思。

operationQueue.addOperations(tasks, waitUntilFinished: false)
```

最后这个是调度组。唯一的区别是，它更容易编写复杂的任务。

## 调度源（DispatchSource）

> DispatchSource用于检测文件和文件夹的变化。根据我们的需要，它有许多变化。我将只在下面展示一个例子。

Reviewer 2: ```suggestion
DispatchSource 用于检测文件和文件夹的变化。根据我们的需要，它有许多变化。我将只在下面展示一个例子。
```

Reviewer: ```suggestion
DispatchSource 用于检测文件和文件夹的变化。根据我们的需要，它有许多变化。我在下面将只展示一个例子。
```

}
```

千万不要在主队列上调用同步任务，这将导致死锁。

## 主线程检查器（Main Thread Checker）

有一种方法可以得到一个警告，说我们做错了什么。这是一个非常有用的选项，我建议使用它。它可以很容易地抓住一些不需要的问题。

> 如果你在目标上打开并编辑下一张图片上的方案，打开主线程检查器，那么当我们在后台做一些UI更新时，运行时的这个选项会通知我们。请看下图中的紫色通知。

Reviewer 2: ```suggestion
如果你在目标上打开并编辑下一张图片上的方案，打开主线程检查器，那么当我们在后台做一些 UI 更新时，运行时的这个选项会通知我们。请看下图中的紫色通知。
```

![Main Thread Checker](https://cdn-images-1.medium.com/max/2000/1*iIBFEKYQx3iX0eaOWuVwPw.png)

![Main Thread Checker result](https://cdn-images-1.medium.com/max/2416/1*l_UquEAZtmgM7bMSYb2vWg.png)

![Method name where is the issue can be seen](https://cdn-images-1.medium.com/max/2262/1*jQORZnI8tlPbxYRmkuwN6A.png)

> 你也可以在Xcode终端看到什么是错误的。对于新手来说，这可能是一个有点奇怪的信息，但很快你就会习惯它。但是你可以在这一行中看到问题所在的方法的名称。

Reviewer 2: ```suggestion
你也可以在 Xcode 终端中看到问题所在。对于新手来说，这可能是一个有点奇怪的信息，但很快你就会习惯它。你可以在这一行中看到问题所在的方法的名称。
```

`po Thread.main`

> 也许你有过类似的情况--当应用程序崩溃时，在错误日志中你可以看到类似com.alamofire.error.serialization.response的东西。这意味着框架创建了一些自定义线程，这就是标识符。

Reviewer 2: ```suggestion
也许你有过类似的情况——当应用程序崩溃时，在错误日志中你可以看到类似 com.alamofire.error.serialization.response 的东西。这意味着框架创建了一些自定义线程，这就是标识符。
```

## Async / Await

> 随着iOS13和Swift 5.5的推出，人们期待已久的Async / Await被引入。苹果公司很机智的认识到了一个问题，那就是当新的东西被引入时，会有一个很长的缓冲时间，直到它可以在生产中使用，因为我们通常需要支持更多的iOS版本。

Reviewer 2: ```suggestion
在 iOS13 和 Swift 5.5 中，引入了人们期待已久的 Async / Await。苹果公司很机智的认识到了一个问题，那就是当新的东西被引入时，会有一个很长的缓冲时间，直到它可以在生产中使用，因为我们通常需要支持更多的 iOS 版本。
```

Reviewer: ```suggestion
在 iOS13 和 Swift 5.5 中，引入了人们期待已久的 Async / Await。苹果公司很机智的认识到了一个问题，那就是当新的东西被引入时，在它可以在生产中使用前会有一个很长的缓冲时间，因为我们通常需要支持更多的 iOS 版本。
```

}
```

这里有一些值得一提的代码：

* `Task.isCancelled`
* Task.init(priority: .background) {}
* Task.detached(priority: .userInitiated) {}
* `Task.cancel()`

> 我想强调一下TaskGroup。这是Awaiting/Async世界中的 "DispatchGroup"。我发现Paul Hudson在这方面有一个非常好的例子[链接](https://www.hackingwithswift.com/quick-start/concurrency/how-to-create-a-task-group-and-add-tasks-to-it)。

Reviewer 2: ```suggestion
我想强调一下 TaskGroup。这是 Awaiting / Async 世界中的 "DispatchGroup"。我发现 Paul Hudson 在这方面有一个非常好的例子[链接](https://www.hackingwithswift.com/quick-start/concurrency/how-to-create-a-task-group-and-add-tasks-to-it)。
```

## Actor

> Actors，是线程安全的引用类型。他们处理数据竞争和并发问题。正如你在下面看到的，访问actor的属性是通过`await`关键字完成的。

Reviewer 2: ```suggestion
Actors 是一个类，它是线程安全的引用类型。它是用来处理数据竞争和并发问题的。正如你在下面看到的，访问 actor 的属性是通过 `await` 关键字完成的。
```

> > “Actors一次只允许一个任务访问其可变状态。 ” — Apple Docs

Reviewer 2: ```suggestion
> “Actors 一次只允许一个任务访问其可变状态。 ” — Apple Docs
```

// Prints "25"
```

## 结论

> 我们已经聊了很多多线程的话题--从UI和后台线程到死锁和DispatchGroup。我相信你现在已经在成为专家的路上了，或者至少为iOS面试中关于多线程主题的问题做好准备。

Reviewer 2: ```suggestion
我们已经聊了很多多线程的话题--从 UI 和后台线程到死锁和 DispatchGroup。我相信你现在已经在成为专家的路上了，或者至少为 iOS 面试中关于多线程主题的问题做好准备。
```
